### ISSUE TITLE
Pass arguments by name

### ISSUE THREAD

**Issue by crutchcorn** [amwmedia/plop#109]
Though I know that argument bypassing was added in v1.9.0, I am looking for functionality that would allow you to have a plopfile like such:
```javascript
    prompts: [{
      type: 'input',
      name: 'name',
      message: 'Please input the name of your program:'
    }, {
      type: 'confirm',
      name: 'server',
      message: 'Do you want to generate a server?'
    }, {
      type: 'confirm',
      name: 'client',
      message: 'Do you want to generate a client?'
    }]
```

And be able to pass in CLI args like:
`plop --plopfile ploppy.js gen -- --name "This is a test" --server`

**Comment by amwmedia** [comment]
this type of approach was certainly considered [when planning the bypass feature](https://github.com/amwmedia/plop/issues/66). It was not the preferred approach because the name of the prompt is not known to anyone by the generator author. The options for `name`, `server`, `client`, etc would need to be documented in some way so that the user of the generator could know how to correctly form their input.

It MIGHT be interesting to see how much work would be involved in supporting this type of input and adding `--help` options to each generator that would print out the names and input expectations. :thinking:

**Comment by amwmedia** [comment]
so with your example code above... a user could type `plop generator-name --help` and get something like...
```
--name         [String] Please input the name of your program:
--server       [Boolean] Do you want to generate a server?
--client       [Boolean] Do you want to generate a client?
```

then the options would be discoverable. Behind the scenes, the plopCLI would just form the correct series of input values, so the end result would be the same regardless of which input method was used

**Comment by crutchcorn** [comment]
While I understand the design concerns where documentation remains an issue, I'd argue that the same needs to be done when using the bypass feature, but in a more confusing way. How does someone know the order in which things are defined in the plopfile? 

That being said, I couldn't agree more - automated generation of help documentation based off of docstrings or additional inputs to a prompt (which wasn't mentioned but would allow the tool to get a helpstring without having too much manual manipulation of files) and plopfile inputs would be highly useful.

I'd be more than willing to write the code to add this feature to make a PR (but would like to get your approval and bang out more of the details before doing so) as I'm going to be making my own CLI tool with `node-plop` as a base anyway with similar behavior

I think this shouldn't be a nightmare to implament, and I could start working on something with a syntax like this:
```javascript
{
      type: 'confirm',
      name: 'wantClient',
      message: 'Do you want to generate a client?',
      help: 'Decides whether or not to generate a client'
}
```

But if you'd rather use the existing `message`, I would understand that. I think it might be nice to have an optional `help` field and a fallback of `message`. From there, we could get the input type from the `type` field with the ability to define a custom type for custom types? I'm not 100% sure how we'd handle the typing though. Would we want to allow the creators of actionTypes to put their own string there so they could have something like `[{val: Number}]` or would we rather do something else and force them to fit within predefined types?

**Comment by amwmedia** [comment]
so, the logic behind the existing implementation is that a user of the generator is familiar with the order of the prompts because they are always the same. This logic breaks down with more complicated generators where you have many questions being asked but works really well for the common case of a generator with 1 - 3 prompts.

I completely agree that it makes sense to have an optional `help` property and then fall back to message if it is not defined.

I think an implementation of this could use the existing logic for the bypass functionality and simply fill in `_` (prompt me) for things that are not specified in the middle. So for example if we have a generator that prompts for `name, server, client` (in that order) and the user provides `--name something --client`, the code to pass `["something", "_", "true"]` to the bypass logic. The existing bypass logic knows how to handle the value with the context of the prompt type it matches up with.

I believe most if not all of the changes needed would be in the `plop` CLI codebase and not `node-plop`

**Comment by crutchcorn** [comment]
I can understand the logic provided - and think it was some nice syntatical sugar for those using their favorite familiar plop generators. No offense meant when I was raising issues with it - I really truly love and respect the work you've done on this great project.

I was thinking about using the bypass functionality under the hood. No need to fix what ain't broke. 

I totally agree that changes would need to be on the CLI side of things rather than the logic component. Certainly a I/O issue and not a processing issue (I hope that makes any sense for anyone but me). 

Now that I generally have a feel for how you'd like to see these things implamented, I'll start cracking my whip at it this week to get some progress done on it. I hope to be submitting a PR somewhat soon for review

**Comment by amwmedia** [comment]
No offense taken 😁. I provide the backstory for context. I really like the idea of having generator help screens that are programmatically handled. I didn't think of that before, but I think it's pretty awesome

**Comment by crutchcorn** [comment]
I'm mostly done with the impamentations, here's a few things that I've had to make the decision on:

- You are able to combine bypassing by array and by name. (Using my example, `plop generate "This is a name" -- --server`). However, it is important to note that plop will prefer named values over array bypassed ones (`plop generate "Test" -- --name "Hi"` will give you `("Hi", "_", "_")`)
- You will have errors thrown at you when you attempt to pass a value that is unknown (`plop generate -- --test Hi` should output a warning and print the help docs. I've also gone ahead and added this error checking to ensure that the user is not attempting to pass too many variables: `plop generate Hi y y y` will now display an error)

Lastly, I wanted to check up on something - are you sure you wanted to use `"_"` as the delimiter for a blank value? I'd personally much rather see `undefined` as the delimiter for blank options as `"_"` would cause an input that's not allowed or handled incorrectly

**Comment by amwmedia** [comment]
Great!

- I like the combination approach and I think it's fine that named values will take priority.
- perfect! I love that we'll be showing the help screen when they provide named values that aren't correct.

I'm not sure I understand what you mean by "_ would cause an input that's not allowed or handled incorrectly". Can you elaborate? I would be worried that if we use "undefined", we're reserving words that can not be used as input to prompt questions (did the user mean the word "undefined" or should I prompt them for their input). "_" is intended to be an analog for the underline spaces on a paper form that you fill out (again, history just for context).

**Comment by crutchcorn** [comment]
All I meant is that if a user wanted to input `"_"`, they wouldn't be able to as we're reserving words (just as you mentioned is the issue with using "undefined"). However, I wasn't suggesting we use the string `"undefined"`, I instead meant that we should use JS' `undefined` type in the array and use `node-plop` to prompt for any `undefined` or `null` inputs to the bypass array (this functionality may already exist; I am not sure). This would not prevent the user from inputting `--name undefined` as `minimist` converts the input to `"undefined"` string and not `undefined` type. I could easily be missing something that would prevent us from doing so - is there anything that comes to mind that would stop us from going down that approach?

**Comment by amwmedia** [comment]
ah, I see. the current bypass functionality supports `_` as a way to say "prompt me for this one". We could process the current bypass input so that `_` is turned into `undefined` behind the scenes. That way the existing behavior of `plop component "name" _ "something else"` would work, but doing something like `plop component -- --name _` would actually result in the value of `name` being the string "_".

I think that made sense :slightly_smiling_face:

**Comment by crutchcorn** [comment]
Looking further into changing `'_'` to `undefined` in the client side itself, I realize that we'd have to make a change to `node-plop` that would be a breaking change. I'm unsure if you're wanting to do such a major breaking change for such a minor feature gain. What I'll go ahead and do is finish up my work here with everything else, not worry about `'_'` passing and then we can create a new bug and discuss the changes done there.

**Comment by amwmedia** [comment]
now available in v2.3.0 👍
